specs/PushNotifications/ToastNotifications-spec.md

```diff
@@ -0,0 +1,551 @@
+        <visual>
+            <binding template = "ToastGeneric">
+                <text>Downloading this week's new music...</text>
+                <progress
+                    title = "{progressTitle}
+                    value = "{progressValue}"
+                    valueStringOverride = "{progressValueString}"
+                    status = "{progressStatus}" />
+            </binding>
+        </visual>
+    </toast>)";
+
+    XmlDocument doc;
+    doc.LoadXml(payload);
+    AppNotification notification(doc);
+    notification.Tag(tag);
+    notification.Group(group);
+
+    // Assign initial values for first notification progress UI
+    AppNotificationProgressData data;
+    data.Title(L"Weekly playlist"); // Binds to {progressTitle} in xml payload
+    data.Value(0.6); // Binds to {progressValue} in xml payload
+    data.ValueStringOverride(L"15/26 songs"); // Binds to {progressValueString} in xml payload
+    data.Status(L"Downloading..."); // Binds to {progressStatus} in xml payload
+    data.SequenceNumber(1);
+
+    notification.Progress(data);
+    AppNotificationManager::Default().Show(notification);
+}
+
+// Send subsequent progress updates
+winrt::Windows::Foundation::IAsyncAction UpdateProgressAsync()
+{
+    winrt::hstring tag = L"weekly-playlist";
+    winrt::hstring group = L"downloads";
+
+    // Assign new values
+    // Note that you only need to assign values that changed. In this example we don't assign progressStatus since we don't need to change it
+    AppNotificationProgressData data;
+    data.Value(0.7); // Binds to {progressValue} in xml payload
+    data.ValueStringOverride(L"18/26 songs"); // Binds to {progressValueString} in xml payload
+    data.SequenceNumber(2);
+
+    auto result = co_await AppNotificationManager::Default().UpdateProgressDataAsync(data, tag, group);
+    if (result == AppNotificationProgressResult::AppNotificationNotFound)
+    {
+        LOG_HR_MSG(E_NOT_SET, "Toast Progress Update Failed since the previous notification update is missing or has been cleared by the user!");
+    }
+}
+```
+# Remarks
+
+## Registration
+
+The developer should always call the Notification Registration API first to register the current process as
+the Activator target.
+
+## Foreground API calls
+
+The developer should always subscribe to AppNotification activation foreground events to intercept notification related actions if
+the app happens to be running in the foreground.
+
+## Manifest Registration
+
+For MSIX, the COM activator GUID and the exe need to be registered in the manifest. The launch args
+would need to be pre-set to a well-known string that defines Activation Triggers.
+
+```xml
+<Extensions>
+  <com:Extension Category="windows.comServer">
+    <com:ComServer>
+      <com:ExeServer Executable="SampleBackgroundApp\SampleBackgroundApp.exe" DisplayName="SampleBackgroundApp" Arguments="----AppNotificationActivationServer">
```

Comment:
We should consider having the notification library parse and validate their Executable path and ensure it matches to the current executable path... Developers could easily mistype their executable, and then be left clueless why activation isn't working if we don't validate things like that.

Reply:
Is it against the rules to specify a different exe? Like maybe you have a light-weight, headless process for handling in-line replies.

Reply:
If the feature team wants to support different EXEs, a parameter in the Register method could be added to specify the different EXE for the 1% of apps that want that additional complexity (and then still ensuring the majority of apps have quality, informative errors... I have prior experience from developers, getting your EXE path incorrect was one of the biggest stumbling blocks developers had when trying to use this before I started automatically validating that in the Toolkit library).

Reply:
I think you're asking for two things:

1. Validation of the path. Seems reasonable.
2. Enforcement of "path must match current running image (or whatever)". Seems unnecessarily restrictive.

Reply:
For now, we are going to enforce that the COM server path in the manifest matches the current process. I do understand that we want to give the developer a way to override this in the future by perhaps including a Process.exe overload. For example,  Register(String processNameExe). We will evaluate as we keep getting feedback.

Reply:
Your feedback has been incorporated via a Simple AppNotificationManager::Register() implementation :)